### PR TITLE
Eventlog implementation

### PR DESCRIPTION
**Pull request by acn-dgopa**
## Description

- Implented eventlog for authorization call
    - Added authorization event model 
    - Added events queue client to help queue the event to the queue storage
    - Added helper class to map the context request and response to the authorization event model
    - Added event log service. The controller calls the eventlog service to queue the event to the storage
    - Added some new unit tests and added assertions to existing unit tests

## Related Issue(s)/PR's
- https://github.com/Altinn/altinn-authentication/pull/336
- https://github.com/Altinn/altinn-auth-audit-log/pull/39

## Verification
- [ ] **Your** code builds clean without any errors or warnings
- [ ] Manual testing done (required)
- [ ] Relevant automated test added (if you find this hard, leave it and we'll help out)
- [ ] All tests run green

## Documentation
- [ ] User documentation is updated with a separate linked PR in [altinn-studio-docs.](https://github.com/Altinn/altinn-studio-docs) (if applicable)

### CODE REVIEW

File: src/Authorization/Controllers/DecisionController.cs [hunk 1]
```diff
@@ -209,6 +215,7 @@ private async Task<XacmlContextResponse> Authorize(XacmlContextRequest decisionR
         {
             decisionRequest = await this._contextHandler.Enrich(decisionRequest);
 
+            EventLogHelper.CreateAuthorizationEvent(_featureManager, _eventLog, decisionRequest, HttpContext);
```
Comment (TheTechArch) [inline src/Authorization/Controllers/DecisionController.cs:218]: We should log this after we have the result so the decision result is included.

File: src/Authorization/appsettings.json [hunk 2]
```diff
@@ -43,5 +43,12 @@
     "ApiRegisterEndpoint": "http://localhost:5101/register/api/v1/",
     "ApiStorageEndpoint": "http://localhost:5010/storage/api/v1/",
     "ApiResourceRegistryEndpoint": "http://localhost:5101/resourceregistry/api/v1/"
+  },
+  "QueueStorageSettings": {
+    "EventLogConnectionString": "DefaultEndpointsProtocol=http;AccountName=devstoreaccount1;AccountKey=Eby8vdM02xNOcqFlqUwJPLlmEtlCDXJ1OUzFT50uSRZ6IFsuFq2UVErCz4I6tq/K1SZFPTOtr/KBHBeksoGMGw==;QueueEndpoint=http://127.0.0.1:10001/devstoreaccount1",
+    "AuthorizationEventQueueName": "authorizationeventlog"
+  },
+  "FeatureManagement": {
+    "AuditLog": true
```
Comment (jonkjetiloye) [inline src/Authorization/appsettings.json:52]: Should default value in appsettings.json not be false for the feature?
And force activation through values.yaml per environment?

File: src/Authorization/Models/ContextRequest.cs [hunk 3]
```diff
@@ -0,0 +1,90 @@
+﻿using System.Collections.Generic;
+using System.ComponentModel.DataAnnotations;
+
+namespace Altinn.Platform.Authorization.Models
+{
+    /// <summary>
+    /// This model describes an authorization event. An authorization event is an action triggered when a user requests access to an operation
+    /// </summary>
+    public class ContextRequest
```
Comment (jonkjetiloye) [inline src/Authorization/Models/ContextRequest.cs:9]: Summary describes this model as: "This model describes an authorization event" then should the model not be called something to this effect? ContextRequest to me sounds like a input request model of some sort.

Comment (jonkjetiloye) [reply src/Authorization/Models/ContextRequest.cs:9]: Oh, it's a copy/paste bug it seems

Comment (jonkjetiloye) [reply src/Authorization/Models/ContextRequest.cs:9]: I still think both the ContextRequest and AccessSubject etc model naming here should have a stronger separation from the actual XacmlContextRequest etc.
And perhaps have their own namespace

File: src/Authorization/Helpers/EventLogHelper.cs [hunk 4]
```diff
@@ -0,0 +1,185 @@
...
+    public static class EventLogHelper
+    {
+        /// <summary>
+        /// Creates an authorization event
+        /// </summary>
+        /// <param name="featureManager">handler for feature manager service</param>
+        /// <param name="eventLog">handler for eventlog service</param>
+        public async static Task CreateAuthorizationEvent(IFeatureManager featureManager, IEventLog eventLog, XacmlContextRequest contextRequest, HttpContext context, XacmlContextResponse contextResponse)
+        {
+            if (await featureManager.IsEnabledAsync(FeatureFlags.AuditLog))
```
Comment (jonkjetiloye) [inline src/Authorization/Helpers/EventLogHelper.cs:31]: Does this belong in a static helper class? To me this is a service implementations responsibility.

File: src/Authorization/Models/AuthorizationEvent.cs [hunk 5]
```diff
@@ -0,0 +1,76 @@
+﻿using System;
+using System.ComponentModel.DataAnnotations;
+
+namespace Altinn.Platform.Authorization.Models
+{
+    /// <summary>
+    /// This model describes an authorization event. An authorization event is an action triggered when a user requests access to an operation
+    /// </summary>
+    public class AuthorizationEvent
```
[the remark this answers is not in the document]
Comment (acn-dgopa) [reply src/Authorization/Models/AuthorizationEvent.cs:9]: yes, they are stored as string in database with the thought that changes in the log producing component will not require constant updating of the log component. But with the lookup in focus, i had asked for a discussion with @lovoll. I hope to adjust the log component based on those inputs.

File: src/Authorization/appsettings.json (hunk 2 above)
Comment (acn-dgopa) [reply src/Authorization/appsettings.json:52]: yeah, that was comitted by mistake. I will fix it

File: src/Authorization/Models/ContextRequest.cs (hunk 3 above)
Comment (acn-dgopa) [reply src/Authorization/Models/ContextRequest.cs:9]: Actually this model is not in use. My last commit failed to remove this. The implementation now is that the enriched XacmlContextRequest is serialized as string and stored as jsonb object in database.

File: src/Authorization/Constants/AltinnXacmlConstants.cs [hunk 6]
```diff
@@ -36,6 +36,11 @@ public static class MatchAttributeIdentifiers
             /// </summary>
             public const string OrgAttribute = "urn:altinn:org";
 
+            /// <summary>
+            /// Org number attribute match indentifier 
+            /// </summary>
+            public const string OrgNumberAttribute = "urn:altinn:orgnr";
```
[the remark this answers is not in the document]
Comment (acn-dgopa) [reply src/Authorization/Constants/AltinnXacmlConstants.cs:42]: I understood from the existing tests that we can expect both <<urn:altinn:organizationnumber>> (ref  AltinnResourceRegistry0003Request.json)and <<urn:altinn:orgnr>> (ref AltinApps0011Request.json).  But i'm not sure why. I assumed different inputs from different applications for a reason?

[the remark this answers is not in the document]
Comment (jonkjetiloye) [reply src/Authorization/Constants/AltinnXacmlConstants.cs:42]: In Xacml attributes I'm only aware of 'urn:altinn:organizationnumber' being used. 'urn:altinn:orgNumber' is used as a claim type in tokens I believe

File: src/Authorization/Models/AuthorizationEvent.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (acn-dgopa) [reply src/Authorization/Models/AuthorizationEvent.cs]: The datatypes are updated after our discussions to help the lookup operations

**Review by jonkjetiloye — COMMENTED** [review]
There is two failing tests at the moment, and some code smells for missing documentation headers and code style which can be reviewed.
For the implementation itself, I have no further comments.

**Pull request merged**